### PR TITLE
fix issue with negative available per validator

### PR DESCRIPTION
**Pull request by joe-bowman**
## 1. Summary
- add guards against negative completion times (should not be possible now, but belt and braces)
- panic (hooks and tx, so safe) on negative allocations
- fixes a test that no longer generates an error, given no satisfiable unbondings
- adds a check on msgResponseType if its not empty string then 46 otherwise 45

## 2.Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## 3. Implementation details

Add guards against invalid completion times.
Panic on negative unbonding allocation.

### CODE REVIEW

**Comment by ajansari95** [comment]
Closes comments from #313 for table based check and closes the PR

File: x/interchainstaking/keeper/redemptions.go [hunk 1]
```diff
@@ -304,6 +308,7 @@ func DetermineAllocationsForUndelegation(currentAllocations map[string]math.Int,
 	for idx := range deltas {
 		deltas[idx].Weight = deltas[idx].Weight.Sub(sdk.NewDecFromInt(maxValue)).Abs()
 		sum = sum.Add(deltas[idx].Weight.TruncateInt().Abs())
+		fmt.Printf("Dropped Val: %s, Weight: %s\n", deltas[idx].ValoperAddress, deltas[idx].Weight.TruncateInt())
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/redemptions.go:311]: Can we remove these extra log lines please

File: x/interchainstaking/keeper/redemptions.go [hunk 2]
```diff
@@ -316,6 +321,9 @@ func DetermineAllocationsForUndelegation(currentAllocations map[string]math.Int,
 			if !ok {
 				availablePerValidator[deltas[idx].ValoperAddress] = sdk.ZeroInt()
 			}
+			fmt.Println("Raw allocation", allocation)
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/redemptions.go:324]: Add her

**Review by joe-bowman — COMMENTED** [review]
Minor bits but otherwise good

File: x/interchainstaking/keeper/redemptions.go (hunk 2 above)
Comment (ajansari95) [reply x/interchainstaking/keeper/redemptions.go:324]: do we want to remove all fmt.Println here?

Comment (ajansari95) [reply x/interchainstaking/keeper/redemptions.go:324]: i can see more

File: app/upgrades.go [hunk 3]
```diff
@@ -198,3 +200,29 @@ func v010400rc6UpgradeHandler(app *Quicksilver) upgradetypes.UpgradeHandler {
...
+				}
+				return false
+			})
+			return false
+		})
+
+		app.InterchainstakingKeeper.IterateRedelegationRecords(ctx, func(_ int64, key []byte, record icstypes.RedelegationRecord) (stop bool) {
+			if record.CompletionTime.Unix() <= 0 {
+				app.InterchainstakingKeeper.Logger(ctx).Info("Removing delegation record", "chainid", record.ChainId, "source", record.Source, "destination", record.Destination, "epoch", record.EpochNumber)
+				app.InterchainstakingKeeper.DeleteRedelegationRecord(ctx, record.ChainId, record.Source, record.Destination, record.EpochNumber)
```
Comment (aljo242) [inline app/upgrades.go:221]: Currently not testing this branch of the `if`.  Otherwise, lgtm with @joe-bowman's comments addressed.

Comment (ajansari95) [reply app/upgrades.go:221]: actually it does the test L153:uprades_test.go - I resued the earlier state to test ; )

**Review by aljo242 — APPROVED** [review]

**Review by ajansari95 — APPROVED** [review]
LGTM

**Pull request merged**